### PR TITLE
Used named scopes outside models

### PR DESCRIPTION
**Pull request by andycroll**
More in a mini ActiveRecord series.

### CODE REVIEW

**Comment by andycroll** [comment]
Don't feel you have to review while we're still sorting the other one!

**Comment by nodunayo** [comment]
👍🏾

—
Nadia Odunayo
nadiaodunayo.com <http://www.nadiaodunayo.com/>

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+layout: article
+category: ruby
+image:
+  base: '2018/only-use-named-scopes-outside-models'
+  alt: 'San Francisco Museum of Modern Art'
+  source: 'https://unsplash.com/photos/RkJF2BMrLJc'
+  credit: 'William Bout'
+---
+
+Last time we discussed using the hash-syntax in your where clauses but in my examples I did something
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *'hash' syntax

* `where`

Comment (nodunayo) [reply _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Is it just me or does the sentence end abruptly...?

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+...only named scopes that you define in your model.
+
+```ruby
+class Post < ApplicationRecord
+  scope :published, -> { where(status: 'published') }
+end
+```
+
+And then use it like:
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *'like so' or 'like this'

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  def index
+    @posts = Post.published
+  end
+end
+```
+
+
+## But why?
+
+This technique is mostly to improve the organisation of your code. It forces you to do two things that can really help keep you productive over time.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *This technique improves the organisation of your code

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+Firstly, naming the concepts you're creating. When you name something you explain it, often for the benefit of “future you”.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *First: naming the concepts you're creating.

*"...future you" or your colleagues

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+Secondly, there's _one_ place to look for all this rather than having arbitrary scopes and conditions defined all over your code base. When you know where all the conditions are defined you'll know where to look when you want to refactor or optimise database performance.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *Second: you have _one_ place to look for all of this. If you define scopes outside of your model, you can end up with arbitrary scopes...over your codebase.

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+
+
+## Why not?
+
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Reword this to read more like: for things like 'limit' and 'order', pagination, etc., no point bothering.

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+There are exceptions to this. I tend not to bother with `#limit` in scopes or simple `#order`s. Or pagination (if you're doing that).
+
+These scopes are less generic than the `#where` conditions and naming them does often enhance understanding. Plus there's no real clarity benefit as the APIs for those classes are pretty straightforward.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Aren't the more generic??

*There's not much to gain in understanding by wrapping them in another named concept (Suggestion...not sure..)

This stuff from "Plus..." isn't clear. APIs for which classes?

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+It's worth bearing in mind that a named scope might still be a good choice for complex ordering or if the sorting is closely related to the conditions specified in a `#where` method.
+
+A good heuristic to using a scope with ordering is 'can I name the concept'.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *'...can I easily name the concept?'

**Review by nodunayo — CHANGES_REQUESTED** [review]

**Comment by nodunayo** [comment]
Decided to have a look anyway... 😁

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  end
+end
+```
+
+
+## But why?
+
+This technique improves the organisation of your code. It forces you to do two things that can really help keep you productive over time.
+
+Firstly: naming the concepts you're creating. When you name something you explain it, often for the benefit of “future you” or your colleagues.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Comma needed after 'something'

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+
+Secondly: you have _one_ place to look for all of this. If you define scopes outside your models you can end up with arbitrary scopes and conditions defined all over your code base. When you know where all the conditions are defined you'll know where to look when you want to refactor or optimise database performance.
+
+
+## Why not?
+
+For scopes involving `#limit`, simple `#order`s or pagination (if you're doing that) there's very little point in bothering to create specific scopes.
+
+Naming non-`#where` scopes often does not enhance understanding. There's no real clarity benefit as the straightforward use of the existing methods gain nothing from wrapping them in a scope.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: I feel like you said the same thing twice...just in a move convoluted way in the second sentence..

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+```ruby
+scope :by_title, -> { sort(:title) } # no benefit?
+scope :recently_updated, -> { sort(updated_at: :desc) } # worth doing
+```
+
+It's worth bearing in mind that a named scope might still be a good choice for complex ordering or if the sorting is closely related to the conditions specified in a `#where` method.
+
+A good heuristic to naming a scope with ordering or limits is 'can I easily name the concept'.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Need a question mark after 'concept'.

**Review by nodunayo — DISMISSED** [review]

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+
+...only named scopes that you define in your model.
+
+```ruby
+class Post < ApplicationRecord
+  scope :published, -> { where(status: 'published') }
+end
+```
+
+And then use it like so:
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *use them

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+This technique improves the organisation of your code. It forces you to do two things that can really help keep you productive over time.
+
+Firstly: naming the concepts you're creating. When you name something, you explain it, often for the benefit of “future you” or your colleagues.
+
+Secondly: you have _one_ place to look for all of this. If you define scopes outside your models you can end up with arbitrary scopes and conditions defined all over your code base. When you know where all the conditions are defined you'll know where to look when you want to refactor or optimise database performance.
+
+
+## Why not?
+
+For scopes involving `#limit`, simple `#order`s or pagination (if you're doing that) there's very little point in bothering to create specific scopes.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Why the "if you're doing that"...why not just take that parenthesised statement out?

Comment (nodunayo) [reply _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: If you do take it out, add a comma after 'pagination'.

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+
+Often, wrapping non-`#where` queries does not enhance understanding. There’s no extra clarity from wrapping simple `ActiveRelation` methods inside a scope.
+
+```ruby
+scope :by_title, -> { sort(:title) } # no benefit?
+scope :recently_updated, -> { sort(updated_at: :desc) } # worth doing
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Hmmm...aren't both of these examples of 'wrapping simple `ActiveRelation` methods inside a scope'?

If you'd named the scope on like 65 as :by_updated_at you could argue that that is also not worth doing, right?

So, I'm not convinced by the paragraph at line 62 (which still seems like saying the same thing twice) and then the examples. I'm not clear on what's "simple" or not..as the thing on line 66 seems simple to me.

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+```
+
+It's worth bearing in mind that a named scope might still be a good choice for complex ordering or if the sorting is closely related to the conditions specified in a `#where` method.
+
+A good heuristic to creating a named scope with ordering or limits is... can I easily name the concept? Is that better than the existing `ActiveRelation` methods?
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *is: can I easily name the concept? Is the result better than the existing...

**Review by nodunayo — DISMISSED** [review]

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md (hunk 14 above)
Comment (andycroll) [reply _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: They are, but one is an example of a 'decent name' and one a 'probably pointless' name. If the name was `by_updated_at` I'd say it was terrible name.

The naming helps explain when you use it. The ordering of `updated_at` means you need extra context, is it 'recently changed' or 'least recently'.

Comment (nodunayo) [reply _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Got it...but maybe you need to reword the preceding paragraph to stress that naming is only beneficial when it helps enhance the meaning...maybe spell out why sort(:title) is pointless (no extra information) sorting by :updated_at needs more clarity.

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+layout: article
+category: ruby
+image:
+  base: '2018/only-use-named-scopes-outside-models'
+  alt: 'San Francisco Museum of Modern Art'
+  source: 'https://unsplash.com/photos/RkJF2BMrLJc'
+  credit: 'William Bout'
+---
+
+Last time we discussed using the hash-syntax in your `#where` methods but in my examples I did something I wouldn't do in my real-life code...
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *'hash syntax' or 'hash-style syntax' (the latter for continuity, but not important)

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+...only named scopes that you define in your model.
+
+```ruby
+class Post < ApplicationRecord
+  scope :published, -> { where(status: 'published') }
+end
+```
+
+And then use them so:
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *like so (I'd say...)

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+  end
+end
+```
+
+
+## But why?
+
+This technique improves the organisation of your code. It forces you to do two things that can really help keep you productive over time.
+
+Firstly: naming the concepts you're creating. When you name something, you explain it, often for the benefit of “future you” or your colleagues.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Maybe there's room for an aside of 'well' here, e.g.:
"When you name something (well), you explain it,..."

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+
+Secondly: you have _one_ place to look for all of this. If you define scopes outside your models you can end up with arbitrary scopes and conditions defined all over your code base. When you know where all the conditions are defined you'll know where to look when you want to refactor or optimise database performance.
+
+
+## Why not?
+
+For scopes involving `#limit`, simple `#order`s or pagination, there's very little point in bothering to create specific scopes.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: Maybe add a short sentence at the end of this paragraph explaining why...?

File: _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md [hunk 20]
```diff
@@ -0,0 +1,72 @@
...
+
+Naming scopes is only beneficial when you gain extra clarity. Sometimes, with non-`#where` queries, there is no enhanced understanding from wrapping simple `ActiveRelation` methods inside a scope.
+
+```ruby
+scope :by_title, -> { sort(:title) } # no benefit?
+scope :by_updated_at, -> { sort(:updated_at) } # terrible name
+scope :recently_updated, -> { sort(updated_at: :desc) } # probably worth doing
+```
+
+It's worth bearing in mind that a named scope might still be a good choice for complex ordering or if the sorting is closely related to the conditions specified in a `#where` method.
```
Comment (nodunayo) [inline _posts/2018/2018-03-04-only-use-named-scopes-ouside-models.md]: *#where query? (As opposed to method)

**Review by nodunayo — APPROVED** [review]
